src/engine/MILPEncoder.cpp

```diff
@@ -0,0 +1,112 @@
+/*********************                                                        */
+/*! \file MILPEncoder.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Wu
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief [[ Add one-line brief description here ]]
+ **
+ ** [[ Add lengthier description here ]]
+ **/
+
+#include "MILPEncoder.h"
+
+MILPEncoder::MILPEncoder( const ITableau &tableau )
+    : _tableau( tableau )
+{}
+
+void MILPEncoder::encodeInputQuery( GurobiWrapper &gurobi,
+                                    const InputQuery &inputQuery )
+{
+    gurobi.reset();
+    // Add variables
+    for ( unsigned var = 0; var < inputQuery.getNumberOfVariables(); var++ )
+    {
+        double lb = _tableau.getLowerBound( var );
+        double ub = _tableau.getUpperBound( var );
+        String varName = Stringf( "x%u", var );
+        gurobi.addVariable( varName, lb, ub );
+        _variableToVariableName[var] = varName;
+    }
+
+    // Add equations
+    for ( const auto &eq : inputQuery.getEquations() )
+    {
+        List<GurobiWrapper::Term> terms;
+        double scalar = eq._scalar;
+        for ( const auto term : eq._addends )
+            terms.append( GurobiWrapper::Term
+                          ( term._coefficient,
+                            Stringf( "x%u", term._variable ) ) );
+        switch ( eq._type )
+        {
+        case Equation::EQ:
+            gurobi.addEqConstraint( terms, scalar );
+            break;
+        case Equation::LE:
+            gurobi.addLeqConstraint( terms, scalar );
+            break;
+        case Equation::GE:
+            gurobi.addGeqConstraint( terms, scalar );
+            break;
+        default:
+            break;
+        }
+    }
+
+    // Add Piecewise-linear Constraints
+    unsigned ind = 0;
+    for ( const auto &plConstraint : inputQuery.getPiecewiseLinearConstraints() )
+    {
+        if ( !plConstraint->isActive() || plConstraint->phaseFixed() )
+            continue;
```

Comment:
This is a little strange. 
For phaseFixed constraint - we should probably encode them also, no? As either the identity or zero function.
for !isActive(), I'm not sure what this means in this context. Assuming MILP encoding is only invoked when Marabou is not really running, this can only happen when the preprocessor disables some constraints?

Reply:
Before encoding the input query, we call applyValidConstraintCaseSplits(), which will update the bounds of the auxiliary variables. These bounds are encoded into gurobi earlier. So this would not be a problem in the case of a ReLU. 
But I agree the way it is now is not general. I've updated that part of the code. Please check again.

Reply:
This feels like setting up a trap for our future selves. Lets at least add an assertion: for each relu that is either inactive or fixed, it must have an auxiliary variable, and that variable's bounds must indicate that the phase has been fixed.

Reply:
added the assertions